accepted/exception-handling.md

```diff
@@ -0,0 +1,240 @@
+  ourselves into a corner. Though, we consider it acceptable if the
+  minimal viable product (MVP) of the implementation does not achieve
+  peak performance for unwinding as long as there is a viable path to
+  improving the performance.
+
+* Compatibility with standard formats. Regardless of unwinding
+  strategy, Wasmtime must continue to work with system profilers which
+  suport DWARF or frame pointers, e.g. `perf`. Furthermore, in
+  Cranelift `cg_clif` should be extended with capabilities to leverage
+  the unwind info support to emit appropriate DWARF, Windows
+  Structured Exception Handling, and so forth. Though, we believe a
+  MVP capable of recovering the `vmctx` register should suffice for
+  critical use-cases.
+
+# Non-requirements
+[non-requirements]: #non-requirements
+
+* No support for the legacy exception handling
+  revision. Justification: the legacy revision is being phased out.
+
+* No support for unwinding across host frames. Justification:
+  unwinding across host frames would require implementing a full DWARF
+  unwinder or equivalent. Note: this just not preclude throwing and
+  catching exceptions when there is a host frame present in the call
+  stack, as we can catch and rethrow exceptions in trampolines on the
+  boundary.
+
+# Proposal sketch
+[proposal]: #proposal
+
+This section outlines the technical details of the proposal for
+implementing support for exception handling in Wasmtime and Cranelift.
+
+## CLIF extensions
+[clif-semantics]: #clif-semantics
+
+We propose two extensions to CLIF.
+
+* A new block type `catch` which is a landing pad for exceptions. This
+  type of block will have exactly one parameter which will be a
+  pointer-sized integer (morally the exception value), e.g. in CLIF
+  syntax
+```clif
+catch block123(v456: i64):
+  ...
+```
+  We disallow regular control flow edges to `catch` blocks. Meaning
+  that neither `jump`, `brif`, or `br_table` are allowed to branch
+  into a `catch` block. Instead, the control flow edges to `catch`
+  must come via a `try_call` instruction.
+
+* A new call instruction `try_call <ok_label>, <exception_label>`,
+  reminiscent of LLVM's `invoke`, where `<ok_label>` must be a regular
+  block and `<exception_label>` must be `catch` block. The block
+  parameters of `<ok_label>` must match the return types of the called
+  function. The semantics is as follows: when `try_call` returns
+  normally, control tranferred to the block named by `<ok_label>`;
+  when `try_call` is unwound, control is tranferred to the block named
+  by `<exception_label>` (note this may happen multiple times in a
+  two-phase exceptions scenario).
+
+* Potentially we may also want a `try_call_indirect <ok_label>,
+  <exception_label>` for indirect calls.
+
+We do not define a CLIF instruction for throwing an
+exception. Instead, exception throwing must be done indirectly via an
+imported function (e.g. a Wasmtime builtin libcall implemented in the
+host/engine).
```

Comment:
I suspect we might actually want an instruction for throwing exceptions: for the overflow-flag ABI approach, we ideally don't want to tail call out to a custom asm-implemented function just to move the exception payload into a particular register, set the overflow flag, and return. We want to do that stuff inline. But I think we would be forced to do that suboptimal approach if we don't have a dedicated instruction.

Of course, when we are doing C++ ABI exceptions and DWARF, we *will* want to call out to `_Unwind_RaiseException` and friends instead.

I think we can choose between the two options in instruction selection with different lowering rules that look at the current calling convention. For the overflow-flags ABI, we'll need a custom calling convention, say `tail-overflow-exceptions` or something instead of our existing `tail` calling convention, and we can check for that or not.

The other option would be a cranelift setting that gets set by the clif producer, similar to how TLS is done. This is a little funky to me tho because we need the new calling convention either way to control whether we do things like clear flags before regular returns or not, and so setting this theoretical option to the overflow-flags version of exceptions can still only work with that special calling convention, so it feels like we'd end up with two knobs to control roughly the same thing.